File: app/src/main/java/com/pin/recommend/MainActivity.java
package com.pin.recommend;

import androidx.appcompat.app.AppCompatActivity;
import androidx.arch.core.util.Function;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.Transformations;

import android.content.Intent;
import android.os.Bundle;

import com.pin.recommend.model.AppDatabase;
import com.pin.recommend.model.dao.RecommendCharacterDao;
import com.pin.recommend.model.entity.Account;
import com.pin.recommend.model.entity.RecommendCharacter;
import com.pin.recommend.model.viewmodel.AccountViewModel;

import java.util.ArrayList;

import static com.pin.recommend.CharacterDetailActivity.INTENT_CHARACTER;


public class MainActivity extends AppCompatActivity {


    public static final String INTENT_ACCOUNT = "com.pin.recommend.MainActivity.INTENT_ACCOUNT";

    private AccountViewModel accountViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final Intent characterListIntent = new Intent(this, CharacterListActivity.class);
        final ArrayList<Intent> intents = new ArrayList<>();
        intents.add(characterListIntent);

        accountViewModel = MyApplication.getAccountViewModel(this);
        LiveData<RecommendCharacter> fixedCharacter = Transformations.switchMap(accountViewModel.getAccount(), new Function<Account, LiveData<RecommendCharacter>>() {
            @Override
            public LiveData<RecommendCharacter> apply(Account input) {
                RecommendCharacterDao characterDao = AppDatabase.getDatabase(MainActivity.this)
                        .recommendCharacterDao();
                Long fixedCharacterId = -1L;
                if(input != null) {
                    fixedCharacterId = input.fixedCharacterId;
                }
                if(fixedCharacterId == null){
                    fixedCharacterId = -1L;
                }

                return characterDao.findTrackedById(fixedCharacterId);
            }
        });
        fixedCharacter.observe(this, new Observer<RecommendCharacter>() {
            @Override
            public void onChanged(RecommendCharacter character) {
                if(character != null){
                    Intent characterDetailIntent = new Intent(MainActivity.this, CharacterDetailActivity.class);
                    characterDetailIntent.putExtra(INTENT_CHARACTER, character);
                    intents.add(characterDetailIntent);
                }
                startActivities(intents.toArray(new Intent[]{}));
                finish();
            }
        });
    }




}

File: app/src/main/java/com/pin/recommend/MainActivity.kt
package com.pin.recommend

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.switchMap
import com.pin.recommend.Constants.PREF_KEY_IS_LOCKED
import com.pin.recommend.model.AppDatabase
import com.pin.recommend.util.PrefUtil
import com.pin.recommend.viewmodel.CharacterDetailsViewModel
import com.pin.util.admob.reward.RemoveAdReward

class MainActivity : AppCompatActivity() {

    private val detailsVM: CharacterDetailsViewModel by lazy {
        ViewModelProvider(this).get(CharacterDetailsViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val characterListIntent = Intent(this, CharacterListActivity::class.java)
        val intents = ArrayList<Intent>()
        intents.add(characterListIntent)

        val fixedCharacter = detailsVM.account.switchMap { input ->
            val characterDao = AppDatabase.getDatabase(this@MainActivity)
                .recommendCharacterDao()
            var fixedCharacterId: Long? = -1L
            if (input != null) {
                fixedCharacterId = input.fixedCharacterId
            }
            if (fixedCharacterId == null) {
                fixedCharacterId = -1L
            }
            characterDao.watchById(fixedCharacterId)
        }

        fixedCharacter.observe(this) { character ->
            if (character != null) {
                val characterDetailIntent =
                    Intent(this@MainActivity, CharacterDetailActivity::class.java)
                characterDetailIntent.putExtra(
                    CharacterDetailActivity.INTENT_CHARACTER,
                    character.id
                )
                intents.add(characterDetailIntent)
            }
            if (isNeedPassCodeConfirmation && PrefUtil.getBoolean(PREF_KEY_IS_LOCKED)) {
                intents.add(PassCodeConfirmationActivity.createIntent(applicationContext))
            }
            isNeedPassCodeConfirmation = false
            startActivities(intents.toArray(arrayOf()))
            finish()
        }

        val reward = RemoveAdReward.getInstance(this)
        //reward.reset()
        reward.checkRewardTime()
        reward.checkNotify()

    }

    private var isNeedPassCodeConfirmation = true
    override fun onTrimMemory(level: Int) {
        super.onTrimMemory(level)
        if (level == TRIM_MEMORY_UI_HIDDEN) {
            isNeedPassCodeConfirmation = true
        }
    }

    companion object {
        const val INTENT_ACCOUNT = "com.pin.recommend.MainActivity.INTENT_ACCOUNT"
    }
}

